### PR TITLE
New aws_region_facts module to be able to get facts from AWS regions

### PR DESCRIPTION
**Pull request by Sodki**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

aws_region_facts

##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 28e39ca42f) last updated 2017/01/25 16:44:10 (GMT +100)
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY

This is a new module that allows to query EC2 and obtain a a list of regions. It's useful if you want to iterate through the regions in search of something specific. I wrote this module because it was something I needed to do.

This implementation uses the same skeleton as `ec2_group_facts`.

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_region_facts.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    # Turn the boto3 result into ansible_friendly_snaked_names
+    snaked_regions = []
+    for region in regions['Regions']:
+        snaked_regions.append(camel_dict_to_snake_dict(region))
+
+    module.exit_json(regions=snaked_regions)
+
+
+from ansible.module_utils.basic import *
+from ansible.module_utils.ec2 import *
```
Comment (j-carl) [inline lib/ansible/modules/cloud/amazon/ec2_region_facts.py]: Please change module imports to the new style. Take a look at [vmware_guest.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest.py) for an example.

File: lib/ansible/modules/cloud/amazon/ec2_region_facts.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+# Gather facts about a single region
+- ec2_region_facts:
+    filters:
+      region-name: eu-west-1
+'''
+
+RETURN = '''
+regions:
+    description: Regions that match the provided filters. Each element consists of a dict with all the information related to that region.
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_region_facts.py]: Please wrap lines at somewhere around 120 chars. You might want to use

```
  description: >
     Regions that match...
```

File: lib/ansible/modules/cloud/amazon/ec2_region_facts.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    try:
+        regions = connection.describe_regions(
+            Filters=ansible_dict_to_boto3_filter_list(sanitized_filters)
+        )
+    except ClientError as e:
+        module.fail_json(msg=e.message, exception=traceback.format_exc(e))
+
+    # Turn the boto3 result into ansible_friendly_snaked_names
+    snaked_regions = []
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_region_facts.py]: this section can be simplified to

```
module.exit_json(regions=[camel_dict_to_snake_dict(region) for region in regions['Regions']])
```

File: lib/ansible/modules/cloud/amazon/ec2_region_facts.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+RETURN = '''
+regions:
+    description: Regions that match the provided filters. Each element consists of a dict with all the information related to that region.
+    type: list
+    sample:
+'''
+
+
+try:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_region_facts.py]: If you import `HAS_BOTO3` from ansible.module_utils.ec2 this section can be

```
try: 
    from botocore.exceptions import ClientError
except ImportError:
    pass # will be detected by imported HAS_BOTO3
```

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
I expect all new AWS modules to pass flake8, and most of my comments will help with that.

**Comment by Sodki** [comment]
@willthames I've renamed the module and implemented your change requests. flake8 is happy on this one as well.

**Comment by Sodki** [comment]
ready_for_review

File: lib/ansible/modules/cloud/amazon/aws_region_facts.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    sample: "[{
+        'endpoint': 'ec2.us-west-1.amazonaws.com',
+        'region_name': 'us-west-1'
+    }]"
+'''
+
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import get_aws_connection_info, ec2_argument_spec, boto3_conn, \
+    ansible_dict_to_boto3_filter_list, camel_dict_to_snake_dict, HAS_BOTO3
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_region_facts.py]: I prefer another from line to a trailing slash.

So line 69 would be
```
from ansible.module_utils.ec2 import ansible_dict_to_boto3_filter_list, camel_dict_to_snake_dict, HAS_BOTO3
```

File: lib/ansible/modules/cloud/amazon/aws_region_facts.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            resource='ec2',
+            region=region,
+            endpoint=ec2_url,
+            **aws_connect_params
+        )
+    else:
+        module.fail_json(msg="region must be specified")
+
+    # Replace filter key underscores with dashes, for compatibility
+    sanitized_filters = module.params.get("filters")
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_region_facts.py]:
```
sanitized_filters = [key.replace("_", "-") for key in module.params.get('filters')]
```

**Comment by willthames** [comment]
shipit

I've made some general comments but they're subjective style rather than essential changes. If you do choose to make those changes, I'll reapply this shipit.

**Comment by Sodki** [comment]
@willthames I've applied the changes you mentioned, makes the code look nicer.

ready_for_review

File: lib/ansible/modules/cloud/amazon/aws_region_facts.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            resource='ec2',
+            region=region,
+            endpoint=ec2_url,
+            **aws_connect_params
+        )
+    else:
+        module.fail_json(msg='region must be specified')
+
+    # Replace filter key underscores with dashes, for compatibility
+    sanitized_filters = {k.replace('_', '-'):v for k,v in module.params.get('filters').items()}
```
Comment (mattclay) [inline lib/ansible/modules/cloud/amazon/aws_region_facts.py]: CI failure due to use of dict comprehension, which isn't supported on python 2.6:

```
2017-02-17 15:38:36 Compiling ./lib/ansible/modules/cloud/amazon/aws_region_facts.py ...
2017-02-17 15:38:36 SyntaxError: ('invalid syntax', ('./lib/ansible/modules/cloud/amazon/aws_region_facts.py', 105, 50, "    sanitized_filters = {k.replace('_', '-'):v for k,v in module.params.get('filters').items()}\n")
```

See here for an alternative: http://stackoverflow.com/questions/21069668/alternative-to-dict-comprehension-prior-to-python-2-7

Comment (Sodki) [reply lib/ansible/modules/cloud/amazon/aws_region_facts.py]: Thank you so much, I've fixed it now.

**Comment by Sodki** [comment]
ready_for_review

**Comment by willthames** [comment]
shipit

**Comment by Sodki** [comment]
ready_for_review

**Comment by mtpereira** [comment]
This looks helpful, ship it! :)

**Review by s-hertel — APPROVED** [review]

**Review by ryansb — APPROVED** [review]

**Pull request merged**